File: static/widget.js
(function () {
  const script = document.currentScript;
  const BOT_ID = script?.getAttribute("data-bot-id");
  const API_BASE = script?.src.replace("/static/widget.js", "") || "";
  if (!BOT_ID) return console.error("[DocBot] data-bot-id is required");

  // persist session across page loads
  const STORAGE_KEY = `docbot_session_${BOT_ID}`;
  let sessionId = sessionStorage.getItem(STORAGE_KEY) || null;

  // ── styles ────────────────────────────────────────────────────────────────
  const style = document.createElement("style");
  style.textContent = `
    #docbot-launcher {
      position: fixed; bottom: 24px; right: 24px; z-index: 9998;
      width: 56px; height: 56px; border-radius: 50%;
      background: #01696f; color: #fff; border: none;
      box-shadow: 0 4px 16px rgba(0,0,0,0.18);
      cursor: pointer; display: flex; align-items: center; justify-content: center;
      transition: transform 180ms ease, background 180ms ease;
    }
    #docbot-launcher:hover { background: #0c4e54; transform: scale(1.08); }
    #docbot-launcher svg { width: 26px; height: 26px; }

    #docbot-frame {
      position: fixed; bottom: 92px; right: 24px; z-index: 9999;
      width: 380px; max-width: calc(100vw - 32px);
      height: 560px; max-height: calc(100vh - 120px);
      border-radius: 16px; overflow: hidden;
      box-shadow: 0 12px 48px rgba(0,0,0,0.22);
      display: none; flex-direction: column;
      background: #f7f6f2; font-family: system-ui, sans-serif;
      border: 1px solid rgba(0,0,0,0.08);
    }
    #docbot-frame.open { display: flex; }

    #docbot-header {
      background: #01696f; color: #fff;
      padding: 14px 16px; font-weight: 600; font-size: 15px;
      display: flex; align-items: center; justify-content: space-between;
      flex-shrink: 0;
    }
    #docbot-header button {
      background: none; border: none; color: #fff; cursor: pointer;
      font-size: 20px; line-height: 1; padding: 0 4px;
    }

    #docbot-messages {
      flex: 1; overflow-y: auto; padding: 14px 12px;
      display: flex; flex-direction: column; gap: 10px;
    }
    .docbot-msg {
      max-width: 80%; padding: 10px 14px; border-radius: 12px;
      font-size: 14px; line-height: 1.5; word-break: break-word;
    }
    .docbot-msg.user {
      align-self: flex-end; background: #01696f; color: #fff;
      border-bottom-right-radius: 4px;
    }
    .docbot-msg.bot {
      align-self: flex-start; background: #fff; color: #28251d;
      border-bottom-left-radius: 4px;
      box-shadow: 0 1px 4px rgba(0,0,0,0.07);
    }
    .docbot-msg.typing { color: #7a7974; font-style: italic; }

    #docbot-input-row {
      display: flex; gap: 8px; padding: 10px 12px;
      border-top: 1px solid rgba(0,0,0,0.08); flex-shrink: 0;
      background: #fff;
    }
    #docbot-input {
      flex: 1; border: 1px solid #d4d1ca; border-radius: 8px;
      padding: 9px 12px; font-size: 14px; outline: none;
      background: #f9f8f5; color: #28251d;
      transition: border-color 180ms ease;
    }
    #docbot-input:focus { border-color: #01696f; }
    #docbot-send {
      background: #01696f; color: #fff; border: none;
      border-radius: 8px; padding: 0 16px; cursor: pointer; font-size: 14px;
      transition: background 180ms ease;
    }
    #docbot-send:hover { background: #0c4e54; }
    #docbot-send:disabled { background: #bab9b4; cursor: not-allowed; }
  `;
  document.head.appendChild(style);

  // ── DOM ───────────────────────────────────────────────────────────────────
  const frame = document.createElement("div");
  frame.id = "docbot-frame";
  frame.innerHTML = `
    <div id="docbot-header">
      <span>💬 Chat</span>
      <button id="docbot-close" aria-label="Close chat">×</button>
    </div>
    <div id="docbot-messages"></div>
    <div id="docbot-input-row">
      <input id="docbot-input" type="text" placeholder="Ask a question…" autocomplete="off" />
      <button id="docbot-send">Send</button>
    </div>
  `;

  const launcher = document.createElement("button");
  launcher.id = "docbot-launcher";
  launcher.setAttribute("aria-label", "Open chat");
  launcher.innerHTML = `<svg viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2">
    <path d="M21 15a2 2 0 0 1-2 2H7l-4 4V5a2 2 0 0 1 2-2h14a2 2 0 0 1 2 2z"/>
  </svg>`;

  document.body.appendChild(frame);
  document.body.appendChild(launcher);

  const messagesEl = frame.querySelector("#docbot-messages");
  const inputEl = frame.querySelector("#docbot-input");
  const sendBtn = frame.querySelector("#docbot-send");

  // ── helpers ───────────────────────────────────────────────────────────────
  function addMessage(text, role) {
    const el = document.createElement("div");
    el.className = `docbot-msg ${role}`;
    el.textContent = text;
    messagesEl.appendChild(el);
    messagesEl.scrollTop = messagesEl.scrollHeight;
    return el;
  }

  // ── toggle ────────────────────────────────────────────────────────────────
  launcher.addEventListener("click", () => {
    frame.classList.toggle("open");
    if (frame.classList.contains("open")) inputEl.focus();
  });
  frame.querySelector("#docbot-close").addEventListener("click", () => {
    frame.classList.remove("open");
  });

  // ── send ──────────────────────────────────────────────────────────────────
  async function sendMessage() {
    const text = inputEl.value.trim();
    if (!text) return;
    inputEl.value = "";
    sendBtn.disabled = true;

    addMessage(text, "user");
    const typing = addMessage("Thinking…", "bot typing");

    try {
      const res = await fetch(`${API_BASE}/public/chat/${BOT_ID}`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ message: text, session_id: sessionId }),
      });
      const data = await res.json();
      if (!res.ok) throw new Error(data.detail || "Error");
      sessionId = data.session_id;
      sessionStorage.setItem(STORAGE_KEY, sessionId);
      typing.className = "docbot-msg bot";
      typing.textContent = data.answer;
    } catch (err) {
      typing.className = "docbot-msg bot";
      typing.textContent = "Sorry, something went wrong. Please try again.";
    } finally {
      sendBtn.disabled = false;
      inputEl.focus();
    }
  }

  sendBtn.addEventListener("click", sendMessage);
  inputEl.addEventListener("keydown", (e) => {
    if (e.key === "Enter" && !e.shiftKey) { e.preventDefault(); sendMessage(); }
  });
})();

File: routers/widget.py
from fastapi import APIRouter, HTTPException
from fastapi.responses import HTMLResponse
from sqlalchemy.ext.asyncio import AsyncSession
from fastapi import Depends
from core.database import get_db
from models.bot import Bot

router = APIRouter(tags=["Widget"])


@router.get("/public/chat-widget/{bot_id}", response_class=HTMLResponse)
async def chat_widget(
    bot_id: str,
    db: AsyncSession = Depends(get_db),
):
    """Serves a minimal embeddable chat UI for a bot. Used by the JS widget snippet."""
    bot = await db.get(Bot, bot_id)
    if not bot or not bot.is_active:
        raise HTTPException(status_code=404, detail="Bot not found")

    # bot.name is safe to embed — it's owner-controlled, not user input
    bot_name = bot.name.replace('"', "&quot;").replace("<", "&lt;").replace(">", "&gt;")

    html = f"""<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>{bot_name}</title>
  <style>
    *, *::before, *::after {{ box-sizing: border-box; margin: 0; padding: 0; }}
    body {{
      font-family: system-ui, sans-serif;
      font-size: 14px;
      display: flex;
      flex-direction: column;
      height: 100dvh;
      background: #f8f8f8;
      color: #1a1a1a;
    }}
    #header {{
      padding: 12px 16px;
      background: #fff;
      border-bottom: 1px solid #e5e5e5;
      font-weight: 600;
      font-size: 15px;
    }}
    #messages {{
      flex: 1;
      overflow-y: auto;
      padding: 16px;
      display: flex;
      flex-direction: column;
      gap: 10px;
    }}
    .msg {{
      max-width: 80%;
      padding: 10px 14px;
      border-radius: 12px;
      line-height: 1.5;
      white-space: pre-wrap;
      word-break: break-word;
    }}
    .msg.user {{
      background: #1a73e8;
      color: #fff;
      align-self: flex-end;
      border-bottom-right-radius: 4px;
    }}
    .msg.bot {{
      background: #fff;
      border: 1px solid #e5e5e5;
      align-self: flex-start;
      border-bottom-left-radius: 4px;
    }}
    .msg.thinking {{
      color: #888;
      font-style: italic;
    }}
    #form {{
      display: flex;
      gap: 8px;
      padding: 12px 16px;
      background: #fff;
      border-top: 1px solid #e5e5e5;
    }}
    #input {{
      flex: 1;
      padding: 9px 12px;
      border: 1px solid #ddd;
      border-radius: 8px;
      font-size: 14px;
      outline: none;
    }}
    #input:focus {{ border-color: #1a73e8; }}
    #send {{
      padding: 9px 18px;
      background: #1a73e8;
      color: #fff;
      border: none;
      border-radius: 8px;
      cursor: pointer;
      font-size: 14px;
      font-weight: 500;
    }}
    #send:disabled {{ opacity: 0.5; cursor: not-allowed; }}
  </style>
</head>
<body>
  <div id="header">{bot_name}</div>
  <div id="messages"></div>
  <form id="form">
    <input id="input" type="text" placeholder="Ask a question…" autocomplete="off" />
    <button id="send" type="submit">Send</button>
  </form>
  <script>
    const BOT_ID = "{bot_id}";
    const API_URL = "/public/chat/" + BOT_ID;
    let sessionId = null;

    const messagesEl = document.getElementById("messages");
    const inputEl = document.getElementById("input");
    const sendEl = document.getElementById("send");
    const formEl = document.getElementById("form");

    function addMessage(role, text) {{
      const div = document.createElement("div");
      div.className = "msg " + role;
      div.textContent = text;
      messagesEl.appendChild(div);
      messagesEl.scrollTop = messagesEl.scrollHeight;
      return div;
    }}

    formEl.addEventListener("submit", async (e) => {{
      e.preventDefault();
      const message = inputEl.value.trim();
      if (!message) return;
      inputEl.value = "";
      sendEl.disabled = true;

      addMessage("user", message);
      const thinking = addMessage("bot thinking", "Thinking…");

      try {{
        const res = await fetch(API_URL, {{
          method: "POST",
          headers: {{ "Content-Type": "application/json" }},
          body: JSON.stringify({{ message, session_id: sessionId }}),
        }});
        const data = await res.json();
        if (!res.ok) throw new Error(data.detail || "Error");
        sessionId = data.session_id;
        thinking.textContent = data.answer;
        thinking.className = "msg bot";
      }} catch (err) {{
        thinking.textContent = "Sorry, something went wrong. Please try again.";
        thinking.className = "msg bot";
      }} finally {{
        sendEl.disabled = false;
        inputEl.focus();
      }}
    }});
  </script>
</body>
</html>"""
    return HTMLResponse(content=html)


@router.get("/public/widget-snippet/{bot_id}", response_class=HTMLResponse)
async def widget_snippet(
    bot_id: str,
    db: AsyncSession = Depends(get_db),
):
    """Returns a copyable HTML snippet to embed the chat widget on any website."""
    bot = await db.get(Bot, bot_id)
    if not bot or not bot.is_active:
        raise HTTPException(status_code=404, detail="Bot not found")

    snippet = f"""<!-- DocBot Widget — paste before </body> -->
<script>
(function() {{
  var iframe = document.createElement('iframe');
  iframe.src = '/public/chat-widget/{bot_id}';
  iframe.style.cssText = 'position:fixed;bottom:80px;right:20px;width:370px;height:500px;border:none;border-radius:12px;box-shadow:0 8px 32px rgba(0,0,0,0.18);z-index:9999;display:none;';
  iframe.id = 'docbot-iframe';
  document.body.appendChild(iframe);

  var btn = document.createElement('button');
  btn.innerHTML = '&#x1F4AC;';
  btn.style.cssText = 'position:fixed;bottom:20px;right:20px;width:52px;height:52px;border-radius:50%;background:#1a73e8;color:#fff;border:none;font-size:22px;cursor:pointer;box-shadow:0 4px 16px rgba(0,0,0,0.18);z-index:9999;';
  btn.onclick = function() {{
    var f = document.getElementById('docbot-iframe');
    f.style.display = f.style.display === 'none' ? 'block' : 'none';
  }};
  document.body.appendChild(btn);
}})();
</script>"""

    return HTMLResponse(content=f"<pre style='font-family:monospace;padding:20px'>{snippet}</pre>")